js/gateio.js

```diff
@@ -2018,61 +2018,123 @@ module.exports = class gateio extends Exchange {
         const market = this.market (symbol);
-        const defaultType = this.safeString2 (this.options, 'createOrder', 'defaultType', 'spot');
-        const marketType = this.safeString (params, 'type', defaultType);
+        const uppercaseType = type.toUpperCase ();
+        const stop_limit = uppercaseType === 'STOP_LOSS_LIMIT' || uppercaseType === 'TAKE_PROFIT_LIMIT';
+        const stop_market = uppercaseType === 'STOP_LOSS_MARKET' || uppercaseType === 'TAKE_PROFIT_MARKET';
+        const limit = uppercaseType === 'LIMIT';
         const contract = market['contract'];
-        const request = this.prepareRequest (market);
-        const reduceOnly = this.safeValue (params, 'reduceOnly');
-        params = this.omit (params, 'reduceOnly');
-        if (reduceOnly !== undefined) {
-            if (!contract) {
-                throw new InvalidOrder (this.id + ' createOrder() does not support reduceOnly for ' + marketType + ' orders, reduceOnly orders are supported for futures and perpetuals only');
+        if (limit || stop_limit) {
+            if (!price) {
+                throw new ArgumentsRequired ('Argument price is required for ' + this.id + '.createOrder for limit orders');
             }
-            request['reduce_only'] = reduceOnly;
-        }
-        if (contract) {
-            if (side === 'sell') {
-                amount = 0 - amount;
+            price = this.priceToPrecision (symbol, price);
+        } else if (!contract) {
+            // Gateio doesn't have market orders for spot
+            throw new InvalidOrder (this.id + ' createOrder() does not support ' + type + ' orders for ' + market['type'] + ' markets');
+        }
+        let request = {};
+        let methodTail = '';
+        amount = this.parseNumber (this.amountToPrecision (symbol, amount));
+        if (stop_limit || stop_market) {
+            const stopPrice = this.safeNumber (params, 'stopPrice');
+            if (stopPrice === undefined) {
+                throw new InvalidOrder (this.id + ' createOrder() requires a stopPrice extra param for a ' + type + ' order');
+            }
+            params = this.omit (params, 'stopPrice');
+            const trigger = {
+                'price': this.priceToPrecision (symbol, stopPrice),
+            };
+            const expiration = this.safeValue (params, 'expiration');
+            if (expiration) {
+                // How long (in seconds) to wait for the condition to be triggered before cancelling the order.
+                trigger['expiration'] = expiration;
+            } else if (!market['contract']) {
+                throw new InvalidOrder (this.id + ' createOrder() requires an expiration extra param for a ' + type + ' order');
             }
-            request['size'] = this.parseNumber (this.amountToPrecision (symbol, amount));
+            const defaultTif = stop_limit ? 'gtc' : 'ioc';
+            if (contract) {
+                trigger['rule'] = side === 'buy' ? 1 : 2;
```

Comment:
This line (and all similar ternary conditionals everywhere) won't transpile to proper Python, cause the transpiler is regex based and it cannot distinguish `side === 'buy' ? 1 : 2` from `'buy' ? 1 : 2`.

Try building it locally with `npm run build` or transpile with `npm run transpile`, then test in Python:

<img width="806" alt="Screenshot 2021-11-06 at 21 32 11" src="https://user-images.githubusercontent.com/1294454/140620190-771df66c-5ca2-455d-84e9-18a4e863293a.png">

We have to wrap all ternary conditionals in brackets:

```JavaScript
(side === 'buy') ? 1 : 2
```